### PR TITLE
Druid gRPC-json server extension

### PR DESCRIPTION
**Pull request by drcrallen**
This is a minimally functional gRPC implementation for Druid. The specific thing this was implemented for was query cancellation. A 0.12 compliant version is used internally in a production environment.

- [x] gRPC server creation
- [x] Retain json polymorphic api spec
- [x] Stand-alone module (designed to be loaded on the brokers only)
- [x] Example clients for `go` and `python`
- [x] Server announcing
- [ ] Druid website docs
- [ ] Figure out how to handle auth

This is opened as a discussion piece since a question was asked in [the dev forum](https://lists.apache.org/thread.html/de28e40a6697e19a6e9e3866e9071c754102eb7cd43f76a1a22fa68b@%3Cdev.druid.apache.org%3E)

I didn't want to invest in making the README stuff until people wanted this to be a full contribution. If it looks interesting let me know and I'll write up more stuff on it.

### CODE REVIEW

File: extensions-core/grpc/src/main/java/org/apache/druid/grpc/QueryServer.java [hunk 1]
```diff
@@ -0,0 +1,511 @@
...
+                    .setName(StringUtils.format(
+                        "%s[%s_%s_%s]",
+                        currThreadName,
+                        query.getType(),
+                        query.getDataSource().getNames(),
+                        id
+                    ));
+              query.getContext().forEach((k, v) -> MDC.put(StringUtils.format("queryContext:%s", k), v.toString()));
+
+              // TODO: some gRPC idiomatic stuff here
```
Comment (drcrallen) [inline extensions-core/grpc/src/main/java/org/apache/druid/grpc/QueryServer.java]: I'm very open to suggestions for better ways to handle this

**Comment by drcrallen** [comment]
For RAT I'm getting

```
Files with unapproved licenses:

  /Users/charles.allen/src/druid/extensions-core/grpc/target/rat.txt
  /Users/charles.allen/src/druid/extensions-core/grpc/target/imports.xml
  /Users/charles.allen/src/druid/extensions-core/grpc/target/checkstyle-cachefile
  /Users/charles.allen/src/druid/extensions-core/grpc/target/checkstyle-result.xml
  /Users/charles.allen/src/druid/extensions-core/grpc/target/pmd.xml
  /Users/charles.allen/src/druid/extensions-core/grpc/target/maven-status/maven-compiler-plugin/testCompile/default-testCompile/inputFiles.lst
  /Users/charles.allen/src/druid/extensions-core/grpc/target/maven-status/maven-compiler-plugin/testCompile/default-testCompile/createdFiles.lst
  /Users/charles.allen/src/druid/extensions-core/grpc/target/maven-status/maven-compiler-plugin/compile/default-compile/inputFiles.lst
  /Users/charles.allen/src/druid/extensions-core/grpc/target/maven-status/maven-compiler-plugin/compile/default-compile/createdFiles.lst
  /Users/charles.allen/src/druid/extensions-core/grpc/target/maven-archiver/pom.properties
  /Users/charles.allen/src/druid/extensions-core/grpc/target/dependency-reduced-pom.xml
  /Users/charles.allen/src/druid/extensions-core/grpc/target/git.version
  /Users/charles.allen/src/druid/extensions-core/grpc/target/.plxarc
```

when I do it locally. But it *thinks* it is excluding `target`

**Comment by drcrallen** [comment]
`useDefaultExcludes` for the rat properties should be excluding the target directory but it is not for some reason.

**Comment by ndolgov** [comment]
@drcrallen For a moment I was scared my little PoC (https://github.com/ndolgov/gruid) was unnecessary.  But then I noticed that your extension is not only of production quality but also goes in a different direction. I have borrowed you Guava workaround though.

**Pull request closed without merging**